common/djangoapps/student/management/commands/bulk_change_enrollment.py

```diff
@@ -83,15 +83,27 @@ def handle(self, *args, **options):
         course_key_str = unicode(course_key)
 
         try:
+            course_enrollments = CourseEnrollment.objects.filter(course_id=course_key, mode=from_mode)
+            logger.info(
+                'Moving %d users from %s to %s in course %s.',
+                course_enrollments.count(), from_mode, to_mode, course_key_str
+            )
+            if not commit:
+                logger.info('Dry run, changes have not been saved. Run again with "commit" argument to save changes')
+                raise Exception('The --commit flag was not given; forcing rollback.')
             with transaction.atomic():
-                queryset = CourseEnrollment.objects.filter(course_id=course_key, mode=from_mode)
-                logger.info(
-                    'Moving %d users from %s to %s in course %s.', queryset.count(), from_mode, to_mode, course_key_str
-                )
-                queryset.update(mode=to_mode)
-
-                if not commit:
-                    raise Exception('The --commit flag was not given; forcing rollback.')
-                logger.info('Finished moving users from %s to %s in course %s.', from_mode, to_mode, course_key_str)
+                # Solution 1: With `change_enrollment`, This will also emit event.
+                for enrollment in course_enrollments:
+                    enrollment.change_mode(mode=to_mode)
+
+                # Solution 2: use `update` and use emit event manually.
+                # #####################################################
+
+                # emit_event_for_enrollments = list(course_enrollments)
+                # course_enrollments.update(mode=to_mode)
+                # for enrollment in emit_event_for_enrollments:
+                #     enrollment.emit_event(EVENT_NAME_ENROLLMENT_MODE_CHANGED)
```

Comment:
I'm not sure how transcations are managed when running management commands.

One advantage of solution 2 is that if you pulled the event emission out of the transaction.atomic block and the transaction is rolled back, no events would be emitted, so the event log would remain synchronized with the database.

Currently when we roll back transactions we don't roll back event emissions so that is a source of error and drift. Not a big enough one for us to fix though...